pkg/bpf/map_linux.go

```diff
@@ -245,11 +257,51 @@ func (m *Map) scheduleErrorResolver() {
 // user space in a local cache (map) and will indicate the status of each
 // individual entry.
 func (m *Map) WithCache() *Map {
-	m.cache = map[string]*cacheEntry{}
+	if m.cache == nil {
+		m.cache = map[string]*cacheEntry{}
+	}
+	m.withValueCache = true
 	m.enableSync = true
 	return m
 }
 
+// WithPressureMetricThreshold enables the tracking of a metric that measures
+// the pressure of this map. This metric is only reported if over the
+// threshold.
+func (m *Map) WithPressureMetricThreshold(threshold float64) *Map {
+	// When pressure metric is enabled, we keep track of map keys in cache
+	if m.cache == nil {
+		m.cache = map[string]*cacheEntry{}
+	}
+
+	m.pressureGauge = metrics.NewBPFMapPressureGauge(m.NonPrefixedName(), threshold)
+
+	return m
+}
+
+// WithPressureMetric enables tracking and reporting of this map pressure with
+// threshold 0.
+func (m *Map) WithPressureMetric() *Map {
+	return m.WithPressureMetricThreshold(0.0)
+}
+
+func (m *Map) updatePressureMetric() {
+	if m.pressureGauge == nil {
+		return
+	}
+
+	if !option.Config.MetricsConfig.BPFMapPressure {
+		if !m.withValueCache {
+			m.cache = nil
+		}
+		m.pressureGauge = nil
+		return
+	}
```

Comment:
[0] Why would `m.pressureGauge` be non-nil at this point if `!option.Config.MetricsConfig.BPFMapPressure`? I would have thought that this field should be initialized to `nil` during the very first attempt to configure the gauge rather than checked & possibly disabled any time there is a map update.

I don't think we have any code that might cause `option.Config.MetricsConfig.BPFMapPressure()` to change at runtime so it should be safe to move it there.

Reply:
On one side we cannot check `option.Config.MetricsConfig.BPFMapPressure` on initialization due to the static initialization of the maps when `option.Config` has not been loaded yet.

`m.pressureGauge` is initialized non-nil. While it could be initialized nil I chose to do it similar and inline with `m.cache` which cannot be initialized nil and which actually allows for less checks.

Reply:
Ah, static initialization and argument parsing strikes again. Thanks for the explanation.

This kind of detail would actually be nice to leave as a comment above the option check to ensure that in future we don't just assume we can move it and rely on the results (unless some other sync mechanism is put in place to ensure the options are available).

Reply:
Added comment!

Reply:
Is this PR (https://github.com/cilium/cilium/pull/14607) related? Is this what's meant by "static initialization"?